File: examples/sample.java
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;

public class Highlighter {
    private String language;
    private boolean inBlockComment;

    public Highlighter(String language) {
        this.language = language;
        this.inBlockComment = false;
    }

    public String highlightLine(String line) {
        if (inBlockComment) {
            int endIndex = line.indexOf("*/");
            if (endIndex == -1) {
                return line;
            }
            inBlockComment = false;
            return line.substring(0, endIndex + 2);
        }

        if (line.startsWith("//")) {
            return line;
        }

        if (line.contains("/*")) {
            if (!line.contains("*/")) {
                inBlockComment = true;
            }
            return line;
        }

        return line;
    }

    public String getLanguage() {
        return this.language;
    }

    @Override
    public String toString() {
        return "Highlighter{" + language + "}";
    }

    public static void main(String[] args) {
        Highlighter h = new Highlighter("java");
        String[] lines = {
            "public static void main(String[] args) {",
            "    System.out.println(\"Hello, world!\");",
            "}"
        };

        for (String line : lines) {
            System.out.println(h.highlightLine(line));
        }

        HashMap<String, Integer> map = new HashMap<>();
        map.put("java", 1);
        map.put("kotlin", 2);

        List<String> keys = new ArrayList<>(map.keySet());
        System.out.println("Languages: " + keys);
    }
}

/* Block comment example
   This should be highlighted
*/

File: examples/sample.kt
import java.util.Objects

data class Highlighter(
    val language: String,
    var inBlockComment: Boolean = false
) {
    fun highlightLine(line: String): String {
        if (inBlockComment) {
            val endIndex = line.indexOf("*/")
            if (endIndex == -1) {
                return line
            }
            inBlockComment = false
            return line.substring(0, endIndex + 2)
        }

        if (line.startsWith("//")) {
            return line
        }

        if (line.contains("/*")) {
            if (!line.contains("*/")) {
                inBlockComment = true
            }
            return line
        }

        return line
    }
}

interface SyntaxEngine {
    fun process(input: String): List<String>
}

sealed class Token {
    data class Keyword(val value: String) : Token()
    data class String(val value: String) : Token()
    data class Number(val value: Int) : Token()
    data class Comment(val value: String) : Token()
    data class Plain(val value: String) : Token()
}

fun createEngine(language: String): SyntaxEngine {
    return object : SyntaxEngine {
        override fun process(input: String): List<String> {
            return input.lines()
        }
    }
}

fun main() {
    val highlighter = Highlighter("kotlin")
    val lines = listOf(
        "fun main() {",
        "    println(\"Hello, world!\")",
        "}"
    )

    for (line in lines) {
        println(highlighter.highlightLine(line))
    }

    val tokens = listOf(
        Token.Keyword("fun"),
        Token.Plain(" "),
        Token.Keyword("main"),
        Token.Plain("() {}")
    )
    println(tokens)
}
